Network/TCP.hs

```diff
@@ -236,15 +236,28 @@ openTCPConnection_ uri port stashInput = do
     -- like this as it just does a once-only installation of a shutdown handler to run at program exit,
     -- rather than actually shutting down after the action
     addrinfos <- withSocketsDo $ getAddrInfo (Just $ defaultHints { addrFamily = AF_UNSPEC, addrSocketType = Stream }) (Just fixedUri) (Just . show $ port)
+
+    let
+      connectAddrInfo a = do
+        s <- socket (addrFamily a) Stream defaultProtocol
+        onException (do
+                    setSocketOption s KeepAlive 1
+                    connect s (addrAddress a)
+                    socketConnection_ fixedUri port s stashInput
+                    ) (Network.Socket.close s)
+
+      tryAddrInfos [] = fail "openTCPConnection: failed to connect to all addresses returned by getAddrInfo"
+      tryAddrInfos (h:t) = connectAddrInfo h `catchIO_` tryAddrInfos t
+
     case addrinfos of
         [] -> fail "openTCPConnection: getAddrInfo returned no address information"
-        (a:_) -> do
-                s <- socket (addrFamily a) Stream defaultProtocol
-                onException (do
-                            setSocketOption s KeepAlive 1
-                            connect s (addrAddress a)
-                            socketConnection_ fixedUri port s stashInput
-                            ) (Network.Socket.close s)
+
+        -- single AddrInfo; call connectAddrInfo directly so that specific
+        -- exception is thrown in event of failure
```

Comment:
You could maybe generalise this by having 'tryAddrInfos' leave out the catch if there's only one element in the list, so that the user always sees one of the connection failures. They might also want to know all addresses that were tried, but that would require more effort with wrapping exceptions etc.

Reply:
I prefer to keep all of the dispatch based on number-of-AddrInfos in one place (currently in the `case addrinfos of ...`).
It is possible to move all that logic to `tryAddrInfos`, but then you lose the distinction between "no AddrInfos returned by getAddrInfo` and "There were multiple addresses, but they all failed".  In my PR these cases have different `fail` error messages, and I think that's a good thing.

Reply:
OK - maybe just include all the addresses in the error message for "they all failed" so the user can then debug by trying them explicitly? It's also ok if you want to leave it as is.

Reply:
@hsenag I updated the code to show the tried addresses in the error messages.

Reply:
> I prefer to keep all of the dispatch based on number-of-AddrInfos in one place (currently in the `case addrinfos of ...`).
> It is possible to move all that logic to `tryAddrInfos`, but then you lose the distinction between "no AddrInfos returned by getAddrInfo`and "There were multiple addresses, but they all failed". In my PR these cases have different`fail` error messages, and I think that's a good thing.

I don't quite understand that you mean, but I'll note that `getAddrInfo` never returns an empty list. If there are no addresses it throws and exception, instead.